### PR TITLE
Add function/0 built-in type

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: lib/elixir/lib/kernel/typespec.ex [hunk 1]
```diff
@@ -595,6 +595,10 @@ defmodule Kernel.Typespec do
     Enum.reduce Enum.reverse(args), fn(arg, expr) -> {:|, [line: line], [arg, expr]} end
   end
 
+  defp typespec_to_ast({:type, line, :function, args}) do
+    typespec_to_ast({:type, line, :fun, args})
```
Comment (josevalim) [inline lib/elixir/lib/kernel/typespec.ex]: `function()` is equivalent to `fun()` but not an `atom()`, I don't think we should support it as there is no benefit for having `function()` then, as `fun()` is already used and also shorter.

**Comment by ericmj** [comment]
What's the use case since it seems to be equivalent with `fun`?

**Comment by josevalim** [comment]
Yes, precisely. Apparently it was listed as `function() :: atom()` in the old OTP docs but that's no longer the case.

I am probably responsible for the confusion though, so apologies @eksperimental.

I will close this as I don't think it brings anything over `fun()`.

**Comment by eksperimental** [comment]
@josevalim that's ok. nothing to worry about.
What leaves me thinking is, if we are not going to officially support it, then I think it will be advised to issue a warning when users use it, because by default we are supporting every type that Erlang supports.

**Comment by josevalim** [comment]
@eksperimental if that's the case, we only need the documentation bit. Everything else already works. Can you add that to your existing PR?

**Pull request closed without merging**